spec/index.bs

```diff
@@ -6682,6 +6726,58 @@ interface mixin GPUProgrammablePassEncoder {
+        :: Pipeline to validate |encoder|s bind groups are compatible with.
+    </dl>
+
+    1. For each |stage| in [{{GPUShaderStage/VERTEX}}, {{GPUShaderStage/FRAGMENT}}, {{GPUShaderStage/COMPUTE}}]:
+        1. Let |writableBufferBindings| be an [=ordered set=]&lt;{{GPUBufferBinding}}&gt;.
+        1. Let |writableTextureViews| be an [=ordered set=]&lt;{{GPUTextureView}}&gt;.
+        1. For each pair of ({{GPUIndex32}} |index|, {{GPUBindGroupLayout}} |bindGroupLayout|) in
+            |pipeline|.{{GPUPipelineBase/[[layout]]}}.{{GPUPipelineLayout/[[bindGroupLayouts]]}}:
+            1. Let |bindGroupEntries| be
+                |encoder|.{{GPUProgrammablePassEncoder/[[bind_groups]]}}[|index|].{{GPUBindGroupDescriptor/entries}}.
+            1. Let |bindGroupLayoutEntries| be
+                |bindGroupLayout|.{{GPUBindGroupLayout/[[descriptor]]}}.{{GPUBindGroupLayoutDescriptor/entries}}.
+            1. For each {{GPUBindGroupEntry}} |bindGroupLayoutEntry| in |bindGroupLayoutEntries|
+                for which |bindGroupLayoutEntry|.{{GPUBindGroupLayoutEntry/visibility}} contains |stage|:
+                1. Let |bindGroupEntry| be the {{GPUBindGroupEntry}} in |bindGroupEntries| for which
+                    |bindGroupEntry|.{{GPUBindGroupEntry/binding}} is equal to
+                    |bindGroupLayoutEntry|.{{GPUBindGroupLayoutEntry/binding}}.
+                1. If |bindGroupEntry|.{{GPUBindGroupEntry/resource}} is a {{GPUBufferBinding}}:
+
+                    1. Let {{GPUBufferBinding}} |bufferBinding| be |bindGroupEntry|.{{GPUBindGroupEntry/resource}}.
+                    1. For each {{GPUBufferBinding}} |existingWritableBufferBinding| in |writableBufferBindings|,
+                        if |existingWritableBufferBinding| and |bufferBinding| are [=buffer-binding-aliasing=], return `true`.
+                    1. If |bindGroupLayoutEntry|.{{GPUBindGroupLayoutEntry/buffer}}.{{GPUBufferBindingLayout/type}}
+                        is {{GPUBufferBindingType/"storage"}}:
+                        1. [=set/append|Append=] |bufferBinding| to |writableBufferBindings|.
+
+                    Otherwise, if |bindGroupEntry|.{{GPUBindGroupEntry/resource}} is a {{GPUTextureView}}:
+
+                    1. Let {{GPUTextureView}} |textureView| be |bindGroupEntry|.{{GPUBindGroupEntry/resource}}.
+                    1. If |bindGroupLayoutEntry|.{{GPUBindGroupLayoutEntry/storageTexture}} is `null`, continue.
+                    1. For each {{GPUTextureView}} |existingWritableTextureView| in |writableTextureViews|,
+                        if |existingWritableTextureView| and |textureView| is [=texture-view-aliasing=], return `true`.
+                    1. If |bindGroupLayoutEntry|.{{GPUBindGroupLayoutEntry/storageTexture}}.{{GPUStorageTextureBindingLayout/access}}
+                        is {{GPUStorageTextureAccess/"write-only"}}:
+                        1. [=set/append|Append=] |textureView| to |writableTextureViews|.
```

Comment:
I think this algorithm may be wrong. If I have a sampled binding followed by the write-only binding, on the same subresource, then:
  - this algorithm would first consider the sampled binding. It would find no conflicts, and proceed without adding it to any lists.
  - then the algorithm would consider the write-only binding. It would find no conflicts, since the last one wasn't registered anywhere.

If the same bindings are checked in a different order (i.e. writable first), they would detect the conflict.

Reply:
Ah, right, I struggled a bit with how to make this algorithm work for read-write conflicts and still got it wrong.